USWDS - In-page navigation: Add custom headings attribute

amyleadem opened the pull request:
# Summary
**Added the ability to customize which headings will be pulled into the in-page navigation link list.**
Use the new `data-heading-elements` attribute to designate the heading levels that should be included in the component. 

The headings should be listed with a space in between. Here is a sample implementation:
```html
<aside 
    class="usa-in-page-nav" 
    data-heading-elements="h3 h4"
</aside>
```

If this attribute is not added to the component or has an empty value, it will default to pulling h2 and h3 headers.

## Breaking change

This is not a breaking change.

## Related issue

Closes #5030 

## Related pull requests

[Changelog PR](https://github.com/uswds/uswds-site/pull/2252)

This work originated in https://github.com/uswds/uswds/pull/5034.

## Preview link

[In-page navigation - custom header selector test](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-in-page-nav-header-select/?path=/story/components-in-page-navigation--test-custom-header-selector)

## Problem statement
The in-page navigation component is hard-coded to pull all h2 and h3 headers from the designated main content region. However, there are valid use cases where teams would want to control which headings will be included in their component. 

## Solution
Creating `data-heading-elements` attribute will allow users select which header levels they want to include in the component link list. This attribute can accept multiple values (between h1-h6), each separated by a space. If the attribute is not added or is left blank, the component will pull the default h2 and h3 headers. 

## Testing and review
- Confirm the attribute name is intuitive and follows convention
- Confirm the correct headings are pulled when the `data-heading-elements` attribute is present
- Confirm that bolded text styles are added to the top-level header in the link list. All other headers in the link list should receive a normal font weight.
- Confirm code follows convention
- Confirm the storybook controls make sense

To test in Storybook:
1. Open the[ In-page navigation - test custom header selector page](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-in-page-nav-header-select/?path=/story/components-in-page-navigation--test-custom-header-selector)
1. In the controls, choose different selections in the "Include these headers in link list" control. Reload the page after each selection change.
1. Confirm that only the expected headers are pulled into the in-page navigation link list.
    1. "Default" should pull h2 and h3 headers
    2. "All" should pull h1-h6 headers
    3. "Error" should show an error in the console and not add any headers to the link list
    4. "Empty" should pull h2 and h3 headers
    5. Single header types should pull only the selected header type
6. Confirm that bolded text styles are added to the highest-level header in the link list. All other headers should have a normal font weight.

To test in a project:
1. Install this branch and copy the updated JS into your project
1. Add the following test markup 
    ```html
    <div class="usa-in-page-nav-container">
      <aside class="usa-in-page-nav"
        data-heading-elements="h2">
      </aside>
      <main>
        <h1>H1 heading</h1>
        <h2>H2 heading</h2>
        <h3>H3 heading</h3>
        <h4>H4 heading</h4>
        <h5>H5 heading</h5>
        <h6>H6 heading</h6>
      </main>
    </div>
    ```
1. Change the value of `data-heading-elements` to any mix of headings and confirm that the link list pulls the correct headings.

amyleadem commented on packages/usa-in-page-navigation/src/test/test-patterns/test-custom-content-selector.twig:
```diff
@@ -1,10 +1,5 @@
 <div class="usa-in-page-nav-container">
   <aside class="usa-in-page-nav"
-    data-title-text="On this page"
-    data-title-heading-level="h4"
-    data-scroll-offset="0"
-    data-root-margin="0px 0px 0px 0px"
-    data-threshold="1"
```
Removed these attributes so that the component uses defaults - this should improve clarity in testing.

amyleadem commented on packages/usa-in-page-navigation/src/index.js:
```diff
@@ -42,23 +44,35 @@ const setActive = (el) => {
+  headingTypes.forEach((headingType) => {
+    if (!IN_PAGE_NAV_VALID_HEADINGS.includes(headingType)) {
+      throw new Error(
+        `Invalid heading type: ${headingType}. Please use one or more of the following: ${IN_PAGE_NAV_VALID_HEADINGS}`
```
Should we include a note about not including punctuation in the error message?

amyleadem replied:
@mejiaj 
1. Yes, it means you need to define the attribute with one or more HTML heading elements (h1-h6). I added some clarity to the error message in https://github.com/uswds/uswds/pull/5444/commits/d7d27a661497cfd7741d9ab5e5e0ec56fb0e65b7
2. I added a note to the error message about not using punctuation in the attribute definition.

mahoneycm reviewed (approved):
Loving this change! Great work 👍 

- [x]  Attribute name is intuitive and follows convention
- [x]  Correct headings are included
- [x]  Links are styled appropriately
- [x]  Test various heading levels in sandbox
    - [x]  `h2 h3 h4`
    - [x]  `h2 h4`
    - [x]  `h4 h2`
    - [x]  `h3`
- [x]  Confirmed highest level heading is respected despite order in data attribute
- [x]  JS matches USWDS standard
- [x]  Automated tests pass

mejiaj commented on packages/usa-in-page-navigation/src/index.js:
```diff
@@ -42,23 +44,35 @@ const setActive = (el) => {
+        `Invalid heading type: ${headingType}. Please use one or more of the following: ${IN_PAGE_NAV_VALID_HEADINGS}`
+      );
+    }
+    contentHeadingTypes.push(`${selectedContentRegion} ${headingType}`);
```
Curious as to the need for `selectedContentRegion` here. I see `main h1,main h2,main h3,main h4,main h5,main h6`.


Do we need each heading prefixed with the content region?

amyleadem replied:
Good note. I reformatted this a bit so that we don't have to make another array in [30a3b9e](https://github.com/uswds/uswds/pull/5444/commits/30a3b9e68787790f8b133a194fd3312b52cd5682). It is similar to the approach in [this comment](https://github.com/uswds/uswds/pull/5444#discussion_r1443051902).

mejiaj commented on packages/usa-in-page-navigation/src/index.js:
```diff
@@ -42,23 +44,35 @@ const setActive = (el) => {
+  const headingTypes = selectedHeadingTypes.indexOf(" ")
+    ? selectedHeadingTypes.split(" ")
+    : selectedHeadingTypes;
+  const contentHeadingTypes = [];
```
Alternative approach to looping and pushing headings would be to query them. Simpler, but unsure if I'm missing something from the current approach.

```js
  // Ideally it'd be a root element for `region`.
  const region = document.querySelector(selectedContentRegion);
  const sectionHeadings = region.querySelectorAll(headingTypes);
```

amyleadem replied:
Made some updates in [30a3b9e](https://github.com/uswds/uswds/pull/5444/commits/30a3b9e68787790f8b133a194fd3312b52cd5682). Let me know if you were looking for anything different.

mejiaj commented on packages/usa-in-page-navigation/src/index.js:
```diff
@@ -216,9 +248,11 @@ const createInPageNav = (inPageNavEl) => {
     const anchorTag = document.createElement("a");
     const textContentOfLink = el.textContent;
     const tag = el.tagName.toLowerCase();
+    const topHeadingLevel = getTopLevelHeading(sectionHeadings);
```
Just noting this runs when there's only a single heading passed. This seems okay for the small amount of headers available.

mejiaj reviewed (changes requested):
Works well and the story controls were helpful! I was able to test:

- [x] All available settings
- [x] Changing order of headings `data-heading-selector="h6 h5 h1 h2 h3 h4"`
- [x] Defaults work as expected

Added some suggestions to improve clarity.

amyleadem commented on packages/usa-in-page-navigation/src/styles/_usa-in-page-navigation.scss:
```diff
@@ -123,13 +123,13 @@
-    font-weight: normal;
-  }
+.usa-in-page-nav__item {
+  font-weight: bold;
```
Updated in [03589a1](https://github.com/uswds/uswds/pull/5444/commits/03589a1218c2c16186f2416ff5cb8a064416a14d)

amyleadem commented on packages/usa-in-page-navigation/src/test/test-patterns/test-custom-header-selector.twig:
Added test descriptions in [9de0209](https://github.com/uswds/uswds/pull/5444/commits/9de0209c721aacb33d86373140bbd14a6d927946)

amyleadem commented on packages/usa-in-page-navigation/src/test/test-patterns/test-hidden-headers.twig:
Updated test description in [9de0209](https://github.com/uswds/uswds/pull/5444/commits/9de0209c721aacb33d86373140bbd14a6d927946)

amyleadem commented on packages/usa-in-page-navigation/src/usa-in-page-navigation.stories.js:
```diff
@@ -18,4 +20,26 @@ export const TestCustomContentSelector = TestCustomContentTemplate.bind();
+      "Error in value",
+      "Empty value",
+    ],
+    control: { type: "inline-radio" },
```
Updated control type in [8eabccc](https://github.com/uswds/uswds/pull/5444/commits/8eabccc76cfcec8591c25e40aaf737324aee2d29)

amyleadem commented on packages/usa-in-page-navigation/src/usa-in-page-navigation.stories.js:
```diff
@@ -18,4 +20,26 @@ export const TestCustomContentSelector = TestCustomContentTemplate.bind();
+      "h4",
+      "h5",
+      "h6",
+      "Error in value",
```
Updated name in [8eabccc](https://github.com/uswds/uswds/pull/5444/commits/8eabccc76cfcec8591c25e40aaf737324aee2d29)

amyleadem commented on packages/usa-in-page-navigation/src/usa-in-page-navigation.stories.js:
```diff
@@ -18,4 +20,26 @@ export const TestCustomContentSelector = TestCustomContentTemplate.bind();
+      "h5",
+      "h6",
+      "Error in value",
+      "Empty value",
```
Removed this option in [8eabccc](https://github.com/uswds/uswds/pull/5444/commits/8eabccc76cfcec8591c25e40aaf737324aee2d29) since it is covered by "default".

amyleadem commented on packages/usa-in-page-navigation/src/usa-in-page-navigation.stories.js:
```diff
@@ -18,4 +20,26 @@ export const TestCustomContentSelector = TestCustomContentTemplate.bind();
+
+export const TestCustomHeaderSelector = TestCustomHeaderTemplate.bind();
+TestCustomHeaderSelector.argTypes = {
+  includeHeaders: {
```
Updated name to `headingType` in[8eabccc](https://github.com/uswds/uswds/pull/5444/commits/8eabccc76cfcec8591c25e40aaf737324aee2d29)

amyleadem commented on packages/usa-in-page-navigation/src/index.js:
```diff
@@ -42,23 +44,35 @@ const setActive = (el) => {
- * Return an array of all visible h2 and h3 headings from the designated main content region.
- * These will be added to the component link list.
+ * Return an array of visible headings from the designated content region.
+ * These items will be added to the component link list.
```
Updated in [30a3b9e](https://github.com/uswds/uswds/pull/5444/commits/30a3b9e68787790f8b133a194fd3312b52cd5682)

amyleadem commented on packages/usa-in-page-navigation/src/index.js:
```diff
@@ -42,23 +44,35 @@ const setActive = (el) => {
+const getSectionHeadings = (selectedContentRegion, selectedHeadingTypes) => {
+  const headingTypes = selectedHeadingTypes.indexOf(" ")
+    ? selectedHeadingTypes.split(" ")
+    : selectedHeadingTypes;
```
Broke this into separate functions: `createSectionHeadingsArray` and `getVisibleSectionHeadings` in [30a3b9e](https://github.com/uswds/uswds/pull/5444/commits/30a3b9e68787790f8b133a194fd3312b52cd5682). Let me know if you want any other changes here.

amyleadem commented on packages/usa-in-page-navigation/src/test/test-patterns/test-custom-header-selector.twig:
```diff
@@ -0,0 +1,32 @@
+<div class="usa-in-page-nav-container">
+  <aside class="usa-in-page-nav"
+    {% if includeHeaders == "All" %}
+      data-heading-selector="h1 h2 h3 h4 h5 h6"
```
It does only accept html header elements. I think it could be a good future enhancement to also allow for an option to exclude or include headers with designated selectors, but that is not a part of this work. 

For clarity, I've updated the name to `data-heading-elements`.

mejiaj replied:
New attribute looks good and found no references to the old one.

amyleadem commented on packages/usa-in-page-navigation/src/test/test-patterns/test-hidden-headers.twig:
> [!NOTE]
> Standardized the visual presentation of the expectations/instructions in this test to match the custom header test.

amyleadem commented:
@mejiaj I believe I have addressed all of your questions/comments. Please let me know if you need anything else.

mejiaj reviewed (changes requested):
Looking good. 

Just sound small outstanding items:
1. Note on BEM I missed on first round (sorry)
2. Please add the release note in the PR description with the bold text + description.
3. Old question in previous review.
4. Question on sanitization in JS.

amyleadem commented on packages/usa-in-page-navigation/src/styles/_usa-in-page-navigation.scss:
```diff
@@ -123,13 +123,13 @@
-    font-weight: normal;
-  }
+.usa-in-page-nav__item {
+  @include u-font-weight("bold");
```
Good catch! Updated in [df47eab](https://github.com/uswds/uswds/pull/5444/commits/df47eab4ac20b8d48041ea6cdaf51cefc56bba73)

amyleadem commented on packages/usa-in-page-navigation/src/index.js:
```diff
@@ -188,14 +242,19 @@ const createInPageNav = (inPageNavEl) => {
   }`;
+  const inPageNavHeadingSelector =
+    inPageNavEl.dataset.headingElements || IN_PAGE_NAV_HEADINGS;
 
```
Whoops. Good catch. Sanitizing makes sense to me. Added in [7a32ee8](https://github.com/uswds/uswds/pull/5444/commits/7a32ee833c3cb36aa81a93a289336128a27f3d79)

amyleadem commented:
@mejiaj Thanks for catching these. Here are my responses to your items: 
1. Note on BEM I missed on first round (sorry)
    - Fixed ([related comment](https://github.com/uswds/uswds/pull/5444#discussion_r1446704507))
1. Please add the release note in the PR description with the bold text + description.
    - Added
1. Old question in previous review.
    - Answered question in [this comment](https://github.com/uswds/uswds/pull/5444#discussion_r1446658396)
1. Question on sanitization in JS.
    - Fixed ([related comment](https://github.com/uswds/uswds/pull/5444#discussion_r1446707531))
    
Let me know if you see anything else.

mejiaj commented on packages/usa-in-page-navigation/src/index.js:
```diff
@@ -188,14 +242,20 @@ const createInPageNav = (inPageNavEl) => {
   };
 
-  const sectionHeadings = getSectionHeadings(inPageNavContentSelector);
+  const sectionHeadings = getVisibleSectionHeadings(
+    inPageNavContentSelector,
+    inPageNavHeadingSelector
+  );
   const inPageNav = document.createElement("nav");
```
Should we have a fallback or message if there are zero sectionHeadings? Consider a unit test to check for empty in-page nav. Pseudocode example below.

**Example**
```js
if (sectionHeadsings.length) {
 // Generate nav.
} else {
	// There aren't any matching visible headings.
	// Consider:
	// 1. Default to all headings.
	// 2. Don't create nav and show message that there aren't matching headers.
}
```

---

## Steps to reproduce

1. Go to `usa-in-page-navigation.twig`
2. In line 11 set change to only use h4's `data-heading-elements="h4"`
3. Visit test url `http://localhost:6006/iframe.html?args=&id=components-in-page-navigation--default&viewMode=story`

<img width="1000" alt="image" src="https://github.com/uswds/uswds/assets/3385219/dded69e3-cc27-4d26-a0aa-9883cad2ec1a">

mejiaj replied:
This issue is captured in #5610.

mejiaj reviewed (changes requested):
Looking good, minor note on nonexistent headers.

**Testing items**

- [x] Single header values like `h1`
- [x] Several headers like `h1, h2, h3`
- [ ] ~Specifying headers that don't appear `ex: h6`. Empty nav is generated, see comment below.~

mejiaj reviewed (approved):
Dismissing [previous review](https://github.com/uswds/uswds/pull/5444#pullrequestreview-1837315118) because issue with empty nav is captured in #5610.

thisisdano reviewed (commented):
I'm going to approve this, but I think there's a modest usability issue here since there's no distinction between the "top" heading level, and all the "lower" heading levels. But that could be a issue for another day
